File: apps/synapse/backend/app/models/rules.py
"""
Rule Engine Data Models

Defines the database models for the hierarchical rule system.
Rules are database-driven and support priority-based execution.
"""

import enum
import uuid
from datetime import datetime

from sqlalchemy import JSON, Boolean, Column, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy import Enum as SQLEnum
from sqlalchemy.orm import relationship

from app.core.database import Base


class RuleSource(str, enum.Enum):
    """
    Source hierarchy for rules.

    Priority order (from lowest to highest):
    - FIRM (10): Baseline engineering standards
    - COUNTRY (30): Country-specific codes and standards
    - PROJECT (50): Project-specific requirements
    - CLIENT (100): Client preferences (highest priority, overrides all)
    """

    FIRM = "FIRM"
    COUNTRY = "COUNTRY"
    PROJECT = "PROJECT"
    CLIENT = "CLIENT"


class RuleDiscipline(str, enum.Enum):
    """Engineering disciplines for rules"""

    ELECTRICAL = "ELECTRICAL"
    AUTOMATION = "AUTOMATION"
    MECHANICAL = "MECHANICAL"
    PROCESS = "PROCESS"
    PIPING = "PIPING"
    INSTRUMENTATION = "INSTRUMENTATION"


class RuleActionType(str, enum.Enum):
    """Types of actions a rule can perform"""

    CREATE_CHILD = "CREATE_CHILD"  # Create related asset (e.g., motor for pump)
    CREATE_CABLE = "CREATE_CABLE"  # Create cable connection
    CREATE_RELATIONSHIP = "CREATE_RELATIONSHIP"  # Create edge between nodes
    SET_PROPERTY = "SET_PROPERTY"  # Set/update asset properties
    CREATE_PACKAGE = "CREATE_PACKAGE"  # Group assets into package
    ALLOCATE_IO = "ALLOCATE_IO"  # Allocate IO card/terminal
    VALIDATE = "VALIDATE"  # Validation rule (check constraints)


class RuleValidationStatus(str, enum.Enum):
    """Status of the rule's lifecycle"""

    DRAFT = "DRAFT"  # Newly created, only runs in test mode
    DEV_VALIDATED = "DEV_VALIDATED"  # Tested by engineer, safe for dev environment
    PROD_READY = "PROD_READY"  # Approved for production/final generation
    DEPRECATED = "DEPRECATED"  # Do not use


class RuleDefinition(Base):
    """
    Database-driven rule definition.

    Rules are applied in priority order:
    CLIENT (100) > PROJECT (50) > COUNTRY (30) > FIRM (10)

    Example rule:
        {
            "name": "Create motor for centrifugal pumps",
            "source": "FIRM",
            "priority": 10,
            "discipline": "ELECTRICAL",
            "action_type": "CREATE_CHILD",
            "condition": {
                "asset_type": "PUMP",
                "property_filters": [
                    {"key": "pump_type", "op": "==", "value": "CENTRIFUGAL"}
                ]
            },
            "action": {
                "create_child": {
                    "type": "MOTOR",
                    "naming": "{parent_tag}-M",
                    "relation": "powers",
                    "discipline": "ELECTRICAL",
                    "inherit_properties": ["area", "system", "voltage"],
                    "properties": {
                        "motor_type": "Electric",
                        "enclosure": "TEFC"
                    }
                }
            }
        }
    """

    __tablename__ = "rule_definitions"

    # Primary Key
    id = Column(String, primary_key=True, default=lambda: f"rule-{uuid.uuid4().hex[:12]}")

    # Basic Info
    name = Column(String(255), nullable=False, index=True)
    description = Column(Text, nullable=True)

    # Source & Priority
    source = Column(SQLEnum(RuleSource), nullable=False, index=True)
    source_id = Column(String, nullable=True)  # FK to country/project/client (if applicable)
    priority = Column(Integer, nullable=False, default=10, index=True)

    # Categorization
    discipline = Column(
        String, nullable=True, index=True
    )  # ELECTRICAL, AUTOMATION, MECHANICAL, PROCESS
    category = Column(
        String, nullable=True, index=True
    )  # ASSET_CREATION, CABLE_SIZING, VALIDATION, ELECTRICAL_CODE, etc.
    action_type = Column(SQLEnum(RuleActionType), nullable=False, index=True)

    # Enforcement
    is_enforced = Column(
        Boolean, default=False, index=True
    )  # Non-negotiable rule (e.g., electrical codes)

    # Rule Logic (JSON)
    condition = Column(JSON, nullable=False)
    # Example: {
    #   "asset_type": "PUMP",
    #   "property_filters": [{"key": "pump_type", "op": "==", "value": "CENTRIFUGAL"}]
    # }

    action = Column(JSON, nullable=False)
    # Example: {
    #   "create_child": {
    #     "type": "MOTOR",
    #     "relation": "powers",
    #     "properties": {...}
    #   }
    # }

    # Conflict Tracking
    overrides_rule_id = Column(
        String, ForeignKey("rule_definitions.id"), nullable=True
    )  # Rule this one overrides
    conflicts_with = Column(
        JSON, nullable=True
    )  # List of conflicting rules: [{rule_id: "xxx", reason: "..."}, ...]

    # Metadata
    validation_status = Column(
        SQLEnum(RuleValidationStatus), default=RuleValidationStatus.DRAFT, index=True
    )
    is_active = Column(Boolean, default=True, index=True)
    created_by = Column(String, ForeignKey("users.id"), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    version = Column(Integer, default=1)

    # Execution Stats (denormalized for performance)
    execution_count = Column(Integer, default=0)
    last_executed_at = Column(DateTime, nullable=True)
    success_count = Column(Integer, default=0)
    failure_count = Column(Integer, default=0)

    # Relationships
    executions = relationship("RuleExecution", back_populates="rule", cascade="all, delete-orphan")
    creator = relationship("User", foreign_keys=[created_by])

    def __repr__(self):
        return f"<RuleDefinition {self.name} (priority={self.priority}, active={self.is_active})>"


class RuleExecution(Base):
    """
    Audit trail for rule execution.

    Logs every action taken by every rule for complete transparency.
    This allows engineers to understand what the system did and why.
    """

    __tablename__ = "rule_executions"

    # Primary Key
    id = Column(String, primary_key=True, default=lambda: f"exec-{uuid.uuid4().hex[:12]}")

    # Context
    rule_id = Column(String, ForeignKey("rule_definitions.id"), nullable=False, index=True)
    project_id = Column(String, ForeignKey("projects.id"), nullable=False, index=True)
    asset_id = Column(String, nullable=True, index=True)  # Asset the rule acted on (if applicable)

    # Execution Result
    action_type = Column(String, nullable=False)  # CREATE, LINK, SKIP, ERROR, UPDATE
    action_taken = Column(Text, nullable=False)  # Human-readable description

    # Technical Details
    condition_matched = Column(Boolean, default=False)
    created_entity_id = Column(String, nullable=True)  # ID of created asset/edge/cable
    created_entity_type = Column(String, nullable=True)  # "MOTOR", "CABLE", "EDGE"

    # Performance Tracking
    timestamp = Column(DateTime, default=datetime.utcnow, index=True)
    execution_time_ms = Column(Integer, nullable=True)  # Execution time in milliseconds

    # Error Handling
    error_message = Column(Text, nullable=True)
    stack_trace = Column(Text, nullable=True)

    # Relationships
    rule = relationship("RuleDefinition", back_populates="executions")
    project = relationship("Project")

    def __repr__(self):
        return f"<RuleExecution {self.action_type} at {self.timestamp}>"


# Backward compatibility aliases
ActionType = RuleActionType

File: apps/synapse/frontend/src/types/rules.ts
/**
 * Type definitions for SYNAPSE Rule Engine
 * 
 * This file provides complete type safety for the rule management system.
 */

// ============================================================================
// ENUMS
// ============================================================================

export enum RuleSource {
    FIRM = 'FIRM',
    COUNTRY = 'COUNTRY',
    PROJECT = 'PROJECT',
    CLIENT = 'CLIENT'
}

export enum RuleActionType {
    CREATE_CHILD = 'CREATE_CHILD',
    SET_PROPERTY = 'SET_PROPERTY',
    CREATE_CABLE = 'CREATE_CABLE',
    CREATE_PACKAGE = 'CREATE_PACKAGE',
    ALLOCATE_IO = 'ALLOCATE_IO',
    VALIDATE = 'VALIDATE'
}

export enum RuleValidationStatus {
    DRAFT = 'DRAFT',
    DEV_VALIDATED = 'DEV_VALIDATED',
    PROD_READY = 'PROD_READY',
    DEPRECATED = 'DEPRECATED'
}

// ============================================================================
// CONDITION TYPES
// ============================================================================

export type ComparisonOperator = '==' | '!=' | '>' | '<' | '>=' | '<=' | 'contains' | 'in';

export interface PropertyFilter {
    key: string;
    op: ComparisonOperator;
    value: string | number | boolean | null;
}

export interface RuleCondition {
    asset_type?: string;
    node_type?: string;  // For matching metamodel nodes (AREA, SITE, etc.)
    property_filters?: PropertyFilter[];
    // Add more condition types as needed
}

// ============================================================================
// ACTION TYPES
// ============================================================================

export interface CreateChildAction {
    type: string;
    naming: string;
    relation: string;
    discipline: string;
    semantic_type: string;
    inherit_properties: string[];
    properties: Record<string, unknown>;
}

export interface SetPropertyAction {
    [key: string]: string | number | boolean | null;
}

export interface CreateCableAction {
    from: string;
    to: string;
    cable_tag: string;
    cable_type: string;
    insulation: string;
    sizing_method: string;
    voltage_drop_limit: number;
}

export interface CreatePackageAction {
    package_type: string;
    naming: string;
    includes: string[];
    excludes: string[];
}

export interface AllocateIOAction {
    cabinet: string;
    card_type: string;
    card_model: string;
    auto_expand: boolean;
}

export interface ValidateAction {
    check: 'has_child' | 'has_property' | 'has_cable';
    child_type?: string;
    property_name?: string;
    error_message: string;
    severity: 'ERROR' | 'WARNING' | 'INFO';
}

export interface RuleAction {
    create_child?: CreateChildAction;
    set_property?: SetPropertyAction;
    create_cable?: CreateCableAction;
    create_package?: CreatePackageAction;
    allocate_io?: AllocateIOAction;
    validate?: ValidateAction;
}

// ============================================================================
// RULE DEFINITION
// ============================================================================

export interface Rule {
    id: string;
    name: string;
    description: string;
    source: RuleSource;
    source_id: string | null;
    priority: number;
    discipline: string;
    category: string;
    action_type: RuleActionType;
    condition: RuleCondition;
    action: RuleAction;
    is_active: boolean;
    is_enforced: boolean;
    overrides_rule_id: string | null;
    conflicts_with: string[];
    execution_count: number;
    success_count: number;
    failure_count: number;
    validation_status?: RuleValidationStatus;
    created_at: string;
    updated_at: string;
}

// ============================================================================
// RULE CREATION (without generated fields)
// ============================================================================

export type RuleCreate = Omit<Rule, 'id' | 'created_at' | 'updated_at' | 'execution_count' | 'success_count' | 'failure_count'>;

export type RuleUpdate = Partial<RuleCreate>;

// ============================================================================
// RULE EXECUTION
// ============================================================================

export interface RuleExecutionResult {
    rule_id: string;
    rule_name: string;
    condition_matched: boolean;
    action_taken: string | null;
    created_entity_id: string | null;
    error_message: string | null;
    timestamp: string;
}

export interface RuleExecutionSummary {
    total_rules: number;
    actions_taken: number;
    execution_time_ms: number;
    details: RuleExecutionResult[];
}

// ============================================================================
// SINGLE RULE EXECUTION
// ============================================================================

export interface SingleRuleExecutionRequest {
    asset_ids?: string[];  // Optional: selective application
}

export interface SingleRuleExecutionResult {
    rule_id: string;
    rule_name: string;
    assets_processed: number;
    actions_taken: number;
    execution_time_ms: number;
    results: Array<{
        asset_id: string;
        asset_tag: string;
        condition_matched: boolean;
        action_taken: string;
        created_entity_id: string | null;
    }>;
}

// ============================================================================
// RULE TEST (with mock data)
// ============================================================================

export interface RuleTestRequest {
    mock_asset: {
        tag: string;
        asset_type: string;
        properties: Record<string, unknown>;
    };
}

export interface RuleTestResult {
    rule_id: string;
    condition_matched: boolean;
    would_create: string | null;
    properties_set: Record<string, unknown> | null;
    explanation: string;
}
